File: native/android/core/network/src/main/kotlin/com/securepastebin/core/network/ApiClient.kt
package com.securepastebin.core.network

import org.json.JSONObject
import java.io.BufferedReader
import java.io.ByteArrayOutputStream
import java.io.InputStream
import java.io.InputStreamReader
import java.net.HttpURLConnection
import java.net.URL
import java.net.URLEncoder
import java.util.UUID

/**
 * API contract and implementation for native Android calls to Secure Pastebin API v1.
 */
interface ApiClient {
    suspend fun uploadEncryptedBlob(data: ByteArray, filename: String): UploadResponse
    suspend fun downloadEncryptedBlob(id: String): DownloadResponse
    suspend fun health(): HealthResponse
}

data class UploadResponse(val id: String, val expiresAt: Long)
data class DownloadResponse(val data: ByteArray)
data class HealthResponse(val configured: Boolean, val account: String?)

/**
 * API error model for transport, protocol, and server failures.
 */
sealed class ApiClientException(message: String) : Exception(message) {
    class Network(message: String) : ApiClientException(message)
    class Protocol(message: String) : ApiClientException(message)
    class Server(val statusCode: Int, message: String) : ApiClientException(message)
}

/**
 * HttpURLConnection-based implementation of the native API v1 client.
 */
class HttpApiClient(
    private val baseUrl: String,
    private val connectTimeoutMs: Int = 30_000,
    private val readTimeoutMs: Int = 30_000,
    private val clientPlatform: String = "android",
    private val clientVersion: String = "unknown",
    private val requestIdProvider: () -> String = { UUID.randomUUID().toString() },
    private val defaultHeaders: Map<String, String> = emptyMap(),
) : ApiClient {
    override suspend fun uploadEncryptedBlob(data: ByteArray, filename: String): UploadResponse {
        val boundary = "SecurePastebinBoundary${UUID.randomUUID().toString().replace("-", "")}"
        val multipartBody = buildMultipartUploadBody(
            data = data,
            filename = filename,
            boundary = boundary,
        )

        val json = requestJson(
            method = "POST",
            path = "/api/v1/upload",
            bodyBytes = multipartBody,
            contentType = "multipart/form-data; boundary=$boundary",
        )

        return UploadResponse(
            id = json.getString("id"),
            expiresAt = json.getLong("expiresAt"),
        )
    }

    override suspend fun downloadEncryptedBlob(id: String): DownloadResponse {
        val encodedId = URLEncoder.encode(id, Charsets.UTF_8.name())
        val json = requestJson(
            method = "GET",
            path = "/api/v1/download?id=$encodedId",
            body = null,
        )

        val raw = json.getJSONArray("data")
        val bytes = ByteArray(raw.length())
        for (index in 0 until raw.length()) {
            bytes[index] = raw.getInt(index).toByte()
        }

        return DownloadResponse(data = bytes)
    }

    override suspend fun health(): HealthResponse {
        val json = requestJson(
            method = "GET",
            path = "/api/v1/health",
            body = null,
        )

        return HealthResponse(
            configured = json.getBoolean("configured"),
            account = if (json.isNull("account")) null else json.getString("account"),
        )
    }

    private fun requestJson(method: String, path: String, body: String?): JSONObject {
        return requestJson(
            method = method,
            path = path,
            bodyBytes = body?.toByteArray(Charsets.UTF_8),
            contentType = if (body != null) "application/json" else null,
        )
    }

    private fun requestJson(
        method: String,
        path: String,
        bodyBytes: ByteArray?,
        contentType: String?,
    ): JSONObject {
        val connection = openConnection(path, method, contentType)
        return try {
            if (bodyBytes != null) {
                connection.doOutput = true
                connection.outputStream.use { stream ->
                    stream.write(bodyBytes)
                }
            }

            val status = connection.responseCode
            val responseText = readResponseBody(connection, status)

            if (status !in 200..299) {
                val message = parseErrorMessage(responseText)
                throw ApiClientException.Server(status, message)
            }

            try {
                JSONObject(responseText)
            } catch (error: Exception) {
                throw ApiClientException.Protocol("Invalid JSON response: ${error.message}")
            }
        } catch (error: ApiClientException) {
            throw error
        } catch (error: Exception) {
            throw ApiClientException.Network(error.message ?: "Network request failed")
        } finally {
            connection.disconnect()
        }
    }

    private fun openConnection(path: String, method: String): HttpURLConnection {
        return openConnection(
            path = path,
            method = method,
            contentType = if (method == "POST") "application/json" else null,
        )
    }

    private fun openConnection(
        path: String,
        method: String,
        contentType: String?,
    ): HttpURLConnection {
        val url = URL(joinUrl(path))
        val connection = (url.openConnection() as HttpURLConnection).apply {
            requestMethod = method
            connectTimeout = connectTimeoutMs
            readTimeout = readTimeoutMs
            // Use a broad accept header to avoid HTML-only route gates on some SSR runtimes.
            setRequestProperty("Accept", "*/*")
            setRequestProperty("X-Client-Platform", clientPlatform)
            setRequestProperty("X-Client-Version", clientVersion)
            setRequestProperty("X-Request-Id", requestIdProvider())
            defaultHeaders.forEach { (header, value) ->
                setRequestProperty(header, value)
            }
            if (!contentType.isNullOrEmpty()) {
                setRequestProperty("Content-Type", contentType)
            }
        }
        return connection
    }

    private fun joinUrl(path: String): String {
        val normalizedBase = if (baseUrl.endsWith("/")) baseUrl.dropLast(1) else baseUrl
        val normalizedPath = if (path.startsWith("/")) path else "/$path"
        return normalizedBase + normalizedPath
    }

    private fun readResponseBody(connection: HttpURLConnection, status: Int): String {
        val stream: InputStream? = if (status in 200..299) {
            connection.inputStream
        } else {
            connection.errorStream
        }

        if (stream == null) {
            return ""
        }

        stream.use { input ->
            BufferedReader(InputStreamReader(input, Charsets.UTF_8)).use { reader ->
                return reader.readText()
            }
        }
    }

    private fun parseErrorMessage(raw: String): String {
        if (raw.isBlank()) {
            return "Unknown server error"
        }

        return try {
            JSONObject(raw).optString("error").ifBlank { raw }
        } catch (_: Exception) {
            raw
        }
    }

    internal fun buildMultipartUploadBody(
        data: ByteArray,
        filename: String,
        boundary: String,
    ): ByteArray {
        val safeFilename = sanitizeMultipartValue(filename)
        val out = ByteArrayOutputStream(data.size + 512)

        out.writeAscii("--$boundary$MULTIPART_LINE_BREAK")
        out.writeAscii(
            "Content-Disposition: form-data; name=\"file\"; filename=\"$safeFilename\"$MULTIPART_LINE_BREAK",
        )
        out.writeAscii("Content-Type: application/octet-stream$MULTIPART_LINE_BREAK$MULTIPART_LINE_BREAK")
        out.write(data)
        out.writeAscii(MULTIPART_LINE_BREAK)

        out.writeAscii("--$boundary$MULTIPART_LINE_BREAK")
        out.writeAscii("Content-Disposition: form-data; name=\"filename\"$MULTIPART_LINE_BREAK$MULTIPART_LINE_BREAK")
        out.writeAscii("$safeFilename$MULTIPART_LINE_BREAK")

        out.writeAscii("--$boundary--$MULTIPART_LINE_BREAK")
        return out.toByteArray()
    }

    private fun sanitizeMultipartValue(value: String): String {
        return value
            .replace("\"", "_")
            .replace("\r", " ")
            .replace("\n", " ")
    }

    private fun ByteArrayOutputStream.writeAscii(value: String) {
        write(value.toByteArray(Charsets.UTF_8))
    }

    private companion object {
        const val MULTIPART_LINE_BREAK = "\r\n"
    }
}

File: native/apple/Sources/CoreNetworking/APIClient.swift
/// Networking abstractions and concrete client for Secure Pastebin native API v1.
import Foundation
#if canImport(FoundationNetworking)
import FoundationNetworking
#endif

/// API contract for Secure Pastebin native calls.
public protocol APIClient {
    /// Executes an upload request against `/api/v1/upload`.
    func uploadEncryptedBlob(data: [UInt8], filename: String) async throws -> UploadResponse

    /// Executes a download request against `/api/v1/download`.
    func downloadEncryptedBlob(id: String) async throws -> DownloadResponse

    /// Checks `/api/v1/health`.
    func health() async throws -> HealthResponse
}

/// Upload response payload for `/api/v1/upload`.
public struct UploadResponse: Decodable, Sendable, Equatable {
    public let id: String
    public let expiresAt: Int64

    public init(id: String, expiresAt: Int64) {
        self.id = id
        self.expiresAt = expiresAt
    }
}

/// Download response payload for `/api/v1/download`.
public struct DownloadResponse: Decodable, Sendable, Equatable {
    public let data: [UInt8]

    public init(data: [UInt8]) {
        self.data = data
    }
}

/// Health response payload for `/api/v1/health`.
public struct HealthResponse: Decodable, Sendable, Equatable {
    public let configured: Bool
    public let account: String?

    public init(configured: Bool, account: String?) {
        self.configured = configured
        self.account = account
    }
}

private struct ErrorResponse: Decodable {
    let error: String
}

/// Errors returned by the API client.
public enum APIClientError: Error, Sendable, LocalizedError, Equatable {
    case invalidBaseURL
    case transport(String)
    case invalidHTTPResponse
    case decodeFailed(String)
    case server(statusCode: Int, message: String)

    public var errorDescription: String? {
        switch self {
        case .invalidBaseURL:
            return "Invalid API base URL"
        case .transport(let message):
            return "Network transport failed: \(message)"
        case .invalidHTTPResponse:
            return "Invalid HTTP response"
        case .decodeFailed(let message):
            return "Failed to decode API response: \(message)"
        case .server(let statusCode, let message):
            return "Server error \(statusCode): \(message)"
        }
    }
}

/// Configuration for URLSession-backed API calls.
public struct APIClientConfiguration: Sendable, Equatable {
    public let baseURL: URL
    public let timeoutSeconds: TimeInterval
    public let defaultHeaders: [String: String]

    public init(
        baseURL: URL,
        timeoutSeconds: TimeInterval = 30,
        defaultHeaders: [String: String] = [:]
    ) {
        self.baseURL = baseURL
        self.timeoutSeconds = timeoutSeconds
        self.defaultHeaders = defaultHeaders
    }
}

/// Concrete URLSession implementation of API v1.
public final class URLSessionAPIClient: APIClient {
    private let configuration: APIClientConfiguration
    private let session: URLSession
    private let clientPlatform: String
    private let clientVersion: String
    private let requestIDProvider: @Sendable () -> String
    private let encoder = JSONEncoder()
    private let decoder = JSONDecoder()

    public init(
        configuration: APIClientConfiguration,
        session: URLSession = .shared,
        clientPlatform: String? = nil,
        clientVersion: String = "unknown",
        requestIDProvider: @escaping @Sendable () -> String = { UUID().uuidString }
    ) {
        self.configuration = configuration
        self.session = session
        self.clientPlatform = clientPlatform ?? URLSessionAPIClient.currentPlatformIdentifier()
        self.clientVersion = clientVersion
        self.requestIDProvider = requestIDProvider
    }

    public func uploadEncryptedBlob(data: [UInt8], filename: String) async throws -> UploadResponse {
        let boundary = "Boundary-\(UUID().uuidString)"
        var request = try makeBaseRequest(
            method: "POST",
            path: "/api/v1/upload"
        )
        request.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")
        request.httpBody = buildMultipartUploadBody(
            data: Data(data),
            filename: filename,
            boundary: boundary
        )
        return try await executeRequest(request)
    }

    public func downloadEncryptedBlob(id: String) async throws -> DownloadResponse {
        let escapedId = id.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? id
        return try await performRequest(
            method: "GET",
            path: "/api/v1/download?id=\(escapedId)",
            body: Optional<Int>.none
        )
    }

    public func health() async throws -> HealthResponse {
        try await performRequest(
            method: "GET",
            path: "/api/v1/health",
            body: Optional<Int>.none
        )
    }

    private func performRequest<RequestBody: Encodable, ResponseBody: Decodable>(
        method: String,
        path: String,
        body: RequestBody?
    ) async throws -> ResponseBody {
        var request = try makeBaseRequest(
            method: method,
            path: path
        )

        if let body {
            request.setValue("application/json", forHTTPHeaderField: "Content-Type")
            request.httpBody = try encoder.encode(body)
        }

        return try await executeRequest(request)
    }

    private func makeBaseRequest(method: String, path: String) throws -> URLRequest {
        let url = try makeURL(path: path)
        var request = URLRequest(url: url)
        request.httpMethod = method
        request.timeoutInterval = configuration.timeoutSeconds
        // Use a broad accept header to avoid HTML-only route gates on some SSR runtimes.
        request.setValue("*/*", forHTTPHeaderField: "Accept")
        request.setValue(clientPlatform, forHTTPHeaderField: "X-Client-Platform")
        request.setValue(clientVersion, forHTTPHeaderField: "X-Client-Version")
        request.setValue(requestIDProvider(), forHTTPHeaderField: "X-Request-Id")

        for (header, value) in configuration.defaultHeaders {
            request.setValue(value, forHTTPHeaderField: header)
        }

        return request
    }

    private func executeRequest<ResponseBody: Decodable>(_ request: URLRequest) async throws -> ResponseBody {
        let rawData: Data
        let rawResponse: URLResponse
        do {
            (rawData, rawResponse) = try await session.data(for: request)
        } catch {
            throw APIClientError.transport(error.localizedDescription)
        }

        guard let response = rawResponse as? HTTPURLResponse else {
            throw APIClientError.invalidHTTPResponse
        }

        guard (200...299).contains(response.statusCode) else {
            let message = decodeServerError(data: rawData)
            throw APIClientError.server(statusCode: response.statusCode, message: message)
        }

        do {
            return try decoder.decode(ResponseBody.self, from: rawData)
        } catch {
            throw APIClientError.decodeFailed(error.localizedDescription)
        }
    }

    private func buildMultipartUploadBody(data: Data, filename: String, boundary: String) -> Data {
        var body = Data()
        let lineBreak = "\r\n"
        let safeFilename = sanitizeMultipartValue(filename)

        body.append("--\(boundary)\(lineBreak)")
        body.append("Content-Disposition: form-data; name=\"file\"; filename=\"\(safeFilename)\"\(lineBreak)")
        body.append("Content-Type: application/octet-stream\(lineBreak)\(lineBreak)")
        body.append(data)
        body.append(lineBreak)

        body.append("--\(boundary)\(lineBreak)")
        body.append("Content-Disposition: form-data; name=\"filename\"\(lineBreak)\(lineBreak)")
        body.append("\(safeFilename)\(lineBreak)")

        body.append("--\(boundary)--\(lineBreak)")
        return body
    }

    private func sanitizeMultipartValue(_ value: String) -> String {
        value
            .replacingOccurrences(of: "\"", with: "_")
            .replacingOccurrences(of: "\r", with: " ")
            .replacingOccurrences(of: "\n", with: " ")
    }

    private func makeURL(path: String) throws -> URL {
        guard var components = URLComponents(
            url: configuration.baseURL,
            resolvingAgainstBaseURL: false
        ) else {
            throw APIClientError.invalidBaseURL
        }

        let safePath = path.hasPrefix("/") ? path : "/\(path)"
        let pathAndQuery = safePath.split(separator: "?", maxSplits: 1, omittingEmptySubsequences: false)
        let requestPath = String(pathAndQuery[0])

        let normalizedBasePath = components.path.hasSuffix("/")
            ? String(components.path.dropLast())
            : components.path
        components.path = normalizedBasePath + requestPath

        if pathAndQuery.count > 1 {
            components.percentEncodedQuery = String(pathAndQuery[1])
        }

        guard let url = components.url else {
            throw APIClientError.invalidBaseURL
        }

        return url
    }

    private func decodeServerError(data: Data) -> String {
        if let decoded = try? decoder.decode(ErrorResponse.self, from: data) {
            return decoded.error
        }
        if let text = String(data: data, encoding: .utf8), !text.isEmpty {
            return text
        }
        return "Unknown server error"
    }

    private static func currentPlatformIdentifier() -> String {
        #if os(iOS)
        return "ios"
        #elseif os(macOS)
        return "macos"
        #else
        return "apple"
        #endif
    }
}

private extension Data {
    mutating func append(_ string: String) {
        append(contentsOf: string.utf8)
    }
}
